src/polymer/polymer2-element-scanner.ts

```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
   async scan(
       document: JavaScriptDocument,
       visit: (visitor: Visitor) => Promise<void>): Promise<ScannedElement[]> {
-    const visitor = new ElementVisitor(document);
-    await visit(visitor);
-    return visitor.getRegisteredElements();
-  }
-}
+    const classFinder = new ClassFinder();
+    const elementDefinitionFinder =
+        new CustomElementsDefineCallFinder(document);
+    // Find all classes and all calls to customElements.define()
+    await Promise.all([visit(classFinder), visit(elementDefinitionFinder)]);
+
+    const elementDefinitionsByClassName = new Map<string, ElementDefineCall>();
+    const elementDefinitionsByClassExpression =
+        new Map<estree.ClassExpression, ElementDefineCall>();
+
+    for (const defineCall of elementDefinitionFinder.calls) {
+      if (defineCall.clazz.type === 'MaybeChainedIdentifier') {
+        elementDefinitionsByClassName.set(defineCall.clazz.name, defineCall);
+      } else {
+        elementDefinitionsByClassExpression.set(defineCall.clazz, defineCall);
+      }
+    }
+    // TODO(rictic): emit ElementDefineCallFeatures for define calls that don't
+    //     map to any local classes?
+
+    const customElements: CustomElementDefinition[] = [];
+    const normalClasses = [];
+    for (const clazz of classFinder.classes) {
+      if (clazz.ast.type === 'ClassExpression') {
+        const definition = elementDefinitionsByClassExpression.get(clazz.ast);
+        if (definition) {
+          customElements.push({clazz, definition});
+          continue;
+        }
+      }
+      const definition =
+          elementDefinitionsByClassName.get(clazz.namespacedName!) ||
+          elementDefinitionsByClassName.get(clazz.name!);
+      if (definition) {
+        customElements.push({clazz, definition});
+        continue;
+      }
+      // TODO(rictic): remove @polymerElement here.
+      //     See: https://github.com/Polymer/polymer-analyzer/issues/605
+      if (jsdoc.hasTag(clazz.doc, 'customElement') ||
+          jsdoc.hasTag(clazz.doc, 'polymerElement')) {
+        customElements.push({clazz});
+        continue;
+      }
+      normalClasses.push(clazz);
+    }
 
-class ElementVisitor implements Visitor {
-  private _possibleElements = new Map<string, ScannedElement>();
-  private _registeredButNotFound = new Map<string, TagNameExpression>();
-  private _elements: Set<ScannedElement> = new Set();
-  private _document: JavaScriptDocument;
-  // TODO(rictic): write a WarningFeature. Emit them from this scanner.
-  private _warnings: Warning[] = [];
+    const elementFeatures: ScannedPolymerElement[] = [];
```

Comment:
golf: `const elementFeatures = customElements.map((e) => this._makeElementFeature(e, document));`